netbox/extras/management/commands/runscript.py

```diff
@@ -0,0 +1,160 @@
+        return scripts[module][name]()
+
+    def handle(self, *args, **options):
+        def _run_script():
+            """
+            Core script execution task. We capture this within a subfunction to allow for conditionally wrapping it with
+            the change_logging context manager (which is bypassed if commit == False).
+            """
+            try:
+                with transaction.atomic():
+                    script.output = script.run(data=data, commit=commit)
+                    job_result.set_status(JobResultStatusChoices.STATUS_COMPLETED)
+
+                    if not commit:
+                        raise AbortTransaction()
+
+            except AbortTransaction:
+                script.log_info("Database changes have been reverted automatically.")
+
+            except Exception as e:
+                stacktrace = traceback.format_exc()
+                script.log_failure(
+                    f"An exception occurred: `{type(e).__name__}: {e}`\n```\n{stacktrace}\n```"
+                )
+                script.log_info("Database changes have been reverted due to error.")
+                logger.error(f"Exception raised during script execution: {e}")
+                job_result.set_status(JobResultStatusChoices.STATUS_ERRORED)
+
+            finally:
+                job_result.data = ScriptOutputSerializer(script).data
+                job_result.save()
+
+            logger.info(f"Script completed in {job_result.duration}")
+
+        # Params
+        script = options['script']
+        loglevel = options['loglevel']
+        data = json.loads(options['data'])
+
+        module, name = script.split('.', 1)
+
+        # Take user from command line if provided and exists, other
+        if options['user']:
+            try:
+                user = User.objects.get(username=options['user'])
+            except User.DoesNotExist:
+                user = User.objects.filter(is_superuser=True).order_by('pk')[0]
+        else:
+            user = User.objects.filter(is_superuser=True).order_by('pk')[0]
+
+        # Setup logging to Stdout
+        formatter = logging.Formatter(f'[%(asctime)s][%(levelname)s] - %(message)s')
+        stdouthandler = logging.StreamHandler(sys.stdout)
+        stdouthandler.setLevel(logging.DEBUG)
+        stdouthandler.setFormatter(formatter)
+
+        logger = logging.getLogger(f"netbox.scripts.{module}.{name}")
+        logger.addHandler(stdouthandler)
+
+        if loglevel == 'debug':
+            logger.setLevel(logging.DEBUG)
+        elif loglevel == 'info':
+            logger.setLevel(logging.INFO)
+        elif loglevel == 'warning':
+            logger.setLevel(logging.WARNING)
+        elif loglevel == 'error':
+            logger.setLevel(logging.ERROR)
+        else:
+            logger.setLevel(logging.INFO)
+
+        # Get the script
+        script = self._get_script(module, name)
+        # Parse the parameters
+        form = script.as_form(data, None)
```

Comment:
Why use a form here? It should be possible to just pass the data directly to the script, as we do with the REST API endpoint for scripts. IMO we shouldn't use forms outside of UI views.

Reply:
The reason is a form seems to resolve ModelObjects to their object, passing the raw data does not.

So if a script is expecting a DeviceObject because someone used a ModelVar and we passed the straight PK it wouldn't work.  Perhaps there is a easier way to do this.